{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 取り組み"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ [ ] right_loxpがいまいちな理由を考察する\n",
    "left_loxpに比べてright_loxpのクラスタがはっきりしない\n",
    "\n",
    "### 補正がうまく行っていないことがわかった\n",
    "\n",
    "- `chr2:180,264,007（329）`の欠失はbartcode31にも42にも同程度で見られる\n",
    "    - にもかかわらず補正がされていない\n",
    "    - ClusteringではなくPreprocessingにおける補正の問題である\n",
    "- `chr2:180,263,057（1279）`の挿入も同様にbartcode31にも42にも同程度で見られる\n",
    "\n",
    "補正がうまく行っていないために、right_loxpのクラスタリングがうまく行っていないと考えられる\n",
    "-> `preprocessing`にてデバッグする\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/d/Research/DAJIN2\n"
     ]
    }
   ],
   "source": [
    "# ルートディレクトリをPathに含めるおまじない\n",
    "import sys, os\n",
    "from pathlib import Path\n",
    "if os.getcwd() != \"/mnt/d/Research/DAJIN2\":\n",
    "    parent_path = str(Path(os.path.dirname(os.path.abspath(\"__file__\"))).parent.parent)\n",
    "    sys.path.append(parent_path)\n",
    "    os.chdir(parent_path)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 実験"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing test-ayabe-task1...\n"
     ]
    }
   ],
   "source": [
    "from __future__ import annotations\n",
    "\n",
    "import sys, os\n",
    "from pathlib import Path\n",
    "\n",
    "sys.path.append(\"/mnt/d/Research/DAJIN2\")\n",
    "os.chdir(\"/mnt/d/Research/DAJIN2\")\n",
    "\n",
    "import hashlib\n",
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "from importlib import reload\n",
    "\n",
    "from src.DAJIN2.core import preprocess, classification, clustering, consensus, report\n",
    "from src.DAJIN2.core.clustering import clustering\n",
    "\n",
    "# # * flox insertion\n",
    "SAMPLE, CONTROL, ALLELE, NAME, GENOME, DEBUG, THREADS = (\n",
    "    \"examples/flox-cables2/AyabeTask1/barcode31.fq.gz\",\n",
    "    \"examples/flox-cables2/AyabeTask1/barcode42.fq.gz\",\n",
    "    \"examples/flox-cables2/AyabeTask1/design_cables2.fa\",\n",
    "    \"test-ayabe-task1\",\n",
    "    \"mm10\",\n",
    "    True,\n",
    "    14,\n",
    ")\n",
    "\n",
    "print(f\"processing {NAME}...\")\n",
    "\n",
    "##########################################################\n",
    "# Check inputs\n",
    "##########################################################\n",
    "preprocess.check_inputs.check_files(SAMPLE, CONTROL, ALLELE)\n",
    "TEMPDIR = Path(\"DAJINResults\", \".tempdir\", NAME)\n",
    "IS_CACHE_CONTROL = preprocess.check_inputs.exists_cached_control(CONTROL, TEMPDIR)\n",
    "IS_CACHE_GENOME = preprocess.check_inputs.exists_cached_genome(GENOME, TEMPDIR, IS_CACHE_CONTROL)\n",
    "UCSC_URL, GOLDENPATH_URL = None, None\n",
    "if GENOME and not IS_CACHE_GENOME:\n",
    "    UCSC_URL, GOLDENPATH_URL = preprocess.check_inputs.check_and_fetch_genome(GENOME)\n",
    "\n",
    "##########################################################\n",
    "# Format inputs\n",
    "##########################################################\n",
    "SAMPLE_NAME = preprocess.format_inputs.extract_basename(SAMPLE)\n",
    "CONTROL_NAME = preprocess.format_inputs.extract_basename(CONTROL)\n",
    "FASTA_ALLELES = preprocess.format_inputs.dictionize_allele(ALLELE)\n",
    "\n",
    "preprocess.format_inputs.make_directories(TEMPDIR, SAMPLE_NAME, CONTROL_NAME)\n",
    "\n",
    "if GENOME:\n",
    "    GENOME_COODINATES = preprocess.format_inputs.fetch_coodinate(GENOME, UCSC_URL, FASTA_ALLELES[\"control\"])\n",
    "    CHROME_SIZE = preprocess.format_inputs.fetch_chrom_size(GENOME_COODINATES[\"chr\"], GENOME, GOLDENPATH_URL)\n",
    "    preprocess.format_inputs.cache_coodinates_and_chromsize(TEMPDIR, GENOME, GENOME_COODINATES, CHROME_SIZE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "\n",
    "import random\n",
    "import re\n",
    "from collections import Counter, defaultdict\n",
    "from copy import deepcopy\n",
    "from pathlib import Path\n",
    "import midsv\n",
    "from scipy import stats\n",
    "from scipy.spatial.distance import cosine\n",
    "\n",
    "\n",
    "def set_indexes(sequence: str):\n",
    "    sequence_length = len(sequence)\n",
    "    num_subset = sequence_length % 5\n",
    "    left_idx = 0\n",
    "    right_idx = sequence_length\n",
    "    if num_subset == 1:\n",
    "        left_idx += 1\n",
    "    elif num_subset == 2:\n",
    "        left_idx += 1\n",
    "        right_idx -= 1\n",
    "    elif num_subset == 3:\n",
    "        left_idx += 2\n",
    "        right_idx -= 1\n",
    "    elif num_subset == 4:\n",
    "        left_idx += 2\n",
    "        right_idx -= 2\n",
    "    return left_idx, right_idx\n",
    "\n",
    "\n",
    "def count_indels_5mer(cssplits: list[list[str]], left_idx: int, right_idx: int) -> list[dict]:\n",
    "    transposed = [list(t) for t in zip(*cssplits)]\n",
    "    count_indels_5mer = []\n",
    "    for i in range(left_idx, right_idx, 5):\n",
    "        count = {\"ins\": [1] * 5, \"del\": [1] * 5, \"sub\": [1] * 5}\n",
    "        cssplits_5mer = transposed[i : i + 5]\n",
    "        for j, cs in enumerate(cssplits_5mer):\n",
    "            counter = Counter(cs)\n",
    "            for key, cnt in counter.items():\n",
    "                if key.startswith(\"=\") or key == \"N\" or re.search(r\"a|c|g|t|n\", key):\n",
    "                    continue\n",
    "                if key.startswith(\"+\"):\n",
    "                    count[\"ins\"][j] += cnt\n",
    "                elif key.startswith(\"-\"):\n",
    "                    count[\"del\"][j] += cnt\n",
    "                elif key.startswith(\"*\"):\n",
    "                    count[\"sub\"][j] += cnt\n",
    "        count_indels_5mer.append(count)\n",
    "    return count_indels_5mer\n",
    "\n",
    "\n",
    "def extract_sequence_errors(count_5mer_sample, count_5mer_control, coverage_sample, coverage_control):\n",
    "    sequence_errors = [set() for _ in range(len(count_5mer_sample))]\n",
    "    for i in range(len(sequence_errors)):\n",
    "        for mutation in [\"ins\", \"del\", \"sub\"]:\n",
    "            samp = [c / coverage_sample for c in count_5mer_sample[i][mutation]]\n",
    "            cont = [c / coverage_control for c in count_5mer_control[i][mutation]]\n",
    "            distance = 1 - cosine(samp, cont)\n",
    "            _, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "            if distance > 0.95 and pvalue > 0.05:\n",
    "                sequence_errors[i].add(mutation)\n",
    "    return sequence_errors\n",
    "\n",
    "\n",
    "def replace_errors_to_atmark(cssplits_sample, sequence_errors, left_idx, right_idx):\n",
    "    cssplits_replaced = []\n",
    "    for samp in cssplits_sample:\n",
    "        samp_replaced = deepcopy(samp)\n",
    "        for idx_error, idx_5mer in enumerate(range(left_idx, right_idx, 5)):\n",
    "            samp_5mer = samp[idx_5mer : idx_5mer + 5]\n",
    "            error = sequence_errors[idx_error]\n",
    "            if \"ins\" in error:\n",
    "                samp_5mer = [\"@\" if cs.startswith(\"+\") else cs for cs in samp_5mer]\n",
    "            if \"del\" in error:\n",
    "                samp_5mer = [\"@\" if cs.startswith(\"-\") else cs for cs in samp_5mer]\n",
    "            if \"sub\" in error:\n",
    "                samp_5mer = [\"@\" if cs.startswith(\"*\") else cs for cs in samp_5mer]\n",
    "            samp_replaced[idx_5mer : idx_5mer + 5] = samp_5mer\n",
    "        cssplits_replaced.append(samp_replaced)\n",
    "    return cssplits_replaced\n",
    "\n",
    "\n",
    "def replace_atmark(cssplits: list[list[str]], sequence: str) -> list[list[str]]:\n",
    "    random.seed(1)\n",
    "    cssplits_replaced = deepcopy(cssplits)\n",
    "    sequence_length = len(sequence)\n",
    "    for i in range(1, sequence_length - 1):\n",
    "        cssplits_atmark = defaultdict(str)\n",
    "        cssplits_sampling_key = defaultdict(list)\n",
    "        cssplits_sampling_all = []\n",
    "        flag_all_atmark = True\n",
    "        for idx, cssplit in enumerate(cssplits):\n",
    "            key = \",\".join([cssplit[i - 1], cssplit[i + 1]])\n",
    "            if cssplit[i] == \"@\":\n",
    "                cssplits_atmark[idx] = key\n",
    "            else:\n",
    "                cssplits_sampling_key[key].append(cssplit[i])\n",
    "                cssplits_sampling_all.append(cssplit[i])\n",
    "                flag_all_atmark = False\n",
    "        for idx, key in cssplits_atmark.items():\n",
    "            if flag_all_atmark:\n",
    "                cssplits_replaced[idx][i] = \"N\"\n",
    "            elif cssplits_sampling_key[key]:\n",
    "                cssplits_replaced[idx][i] = random.choice(cssplits_sampling_key[key])\n",
    "            else:\n",
    "                cssplits_replaced[idx][i] = random.choice(cssplits_sampling_all)\n",
    "    for cs in cssplits_replaced:\n",
    "        if cs[0] == \"@\":\n",
    "            cs[0] = \"=\" + sequence[0]\n",
    "        if cs[-1] == \"@\":\n",
    "            cs[-1] = \"=\" + sequence[-1]\n",
    "    return cssplits_replaced\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "# main\n",
    "###############################################################################\n",
    "\n",
    "\n",
    "def execute(TEMPDIR: Path, FASTA_ALLELES: dict[str, str], CONTROL_NAME: str, SAMPLE_NAME: str) -> None:\n",
    "    for allele, sequence in FASTA_ALLELES.items():\n",
    "        midsv_sample = midsv.read_jsonl((Path(TEMPDIR, \"midsv\", f\"{SAMPLE_NAME}_splice_{allele}.jsonl\")))\n",
    "        midsv_control = midsv.read_jsonl((Path(TEMPDIR, \"midsv\", f\"{CONTROL_NAME}_splice_{allele}.jsonl\")))\n",
    "        coverage_sample = len(midsv_sample)\n",
    "        coverage_control = len(midsv_control)\n",
    "        cssplits_sample = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_sample]\n",
    "        cssplits_control = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_control]\n",
    "        left_idx, right_idx = set_indexes(sequence)\n",
    "        count_5mer_sample = count_indels_5mer(cssplits_sample, left_idx, right_idx)\n",
    "        count_5mer_control = count_indels_5mer(cssplits_control, left_idx, right_idx)\n",
    "        sequence_errors = extract_sequence_errors(\n",
    "            count_5mer_sample, count_5mer_control, coverage_sample, coverage_control\n",
    "        )\n",
    "        cssplits_sample_error_replaced = replace_errors_to_atmark(cssplits_sample, sequence_errors, left_idx, right_idx)\n",
    "        cssplits_control_error_replaced = replace_errors_to_atmark(\n",
    "            cssplits_control, sequence_errors, left_idx, right_idx\n",
    "        )\n",
    "        cssplits_sample_atmark_replaced = replace_atmark(cssplits_sample_error_replaced, sequence)\n",
    "        cssplits_control_atmark_replaced = replace_atmark(cssplits_control_error_replaced, sequence)\n",
    "        # Replace CSSPLIT\n",
    "        cssplits_sample_corrected = [\",\".join(cs) for cs in cssplits_sample_atmark_replaced]\n",
    "        cssplits_control_corrected = [\",\".join(cs) for cs in cssplits_control_atmark_replaced]\n",
    "        for i, cssplits in enumerate(cssplits_sample_corrected):\n",
    "            midsv_sample[i][\"CSSPLIT\"] = cssplits\n",
    "        for i, cssplits in enumerate(cssplits_control_corrected):\n",
    "            midsv_control[i][\"CSSPLIT\"] = cssplits\n",
    "        midsv.write_jsonl(midsv_control, Path(TEMPDIR, \"midsv\", f\"{CONTROL_NAME}_splice_{allele}.jsonl\"))\n",
    "        midsv.write_jsonl(midsv_sample, Path(TEMPDIR, \"midsv\", f\"{SAMPLE_NAME}_splice_{allele}.jsonl\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele = \"control\"\n",
    "sv = False\n",
    "sequence = FASTA_ALLELES[allele]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "midsv_sample = midsv.read_jsonl((Path(TEMPDIR, \"midsv\", f\"{SAMPLE_NAME}_splice_{allele}.jsonl\")))\n",
    "midsv_control = midsv.read_jsonl((Path(TEMPDIR, \"midsv\", f\"{CONTROL_NAME}_splice_{allele}.jsonl\")))\n",
    "coverage_sample = len(midsv_sample)\n",
    "coverage_control = len(midsv_control)\n",
    "cssplits_sample = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_sample]\n",
    "cssplits_control = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_control]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "i_del = 329\n",
    "i_ins = 1280"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=====================\n",
      "control @ 329\n",
      "defaultdict(<class 'int'>, {'=C': 824, '-C': 125, 'N': 8, '+A|*CA': 3, '+C|=C': 3, '+C|+C|=C': 4, '+G|*CT': 1})\n",
      "=====================\n",
      "sample @ 329\n",
      "defaultdict(<class 'int'>, {'=C': 831, '+C|=C': 9, '-C': 129, 'N': 10})\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from collections import Counter\n",
    "\n",
    "count = defaultdict(int)\n",
    "for cs in cssplits_control:\n",
    "    count[cs[i_del]] += 1\n",
    "print(\"=====================\")\n",
    "print(f\"control @ {i_del}\")\n",
    "print(count)\n",
    "\n",
    "count = defaultdict(int)\n",
    "for cs in cssplits_sample:\n",
    "    count[cs[i_del]] += 1\n",
    "print(\"=====================\")\n",
    "print(f\"sample @ {i_del}\")\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=====================\n",
      "control @ 1280\n",
      "defaultdict(<class 'int'>, {'=G': 850, '+A|=G': 26, '-G': 25, '+G|=G': 2, 'N': 34, '+A|+G|=G': 19, '+A|+A|=G': 1, '+G|+A|+G|=G': 1, '+G|+A|=G': 2, '+G|+T|=G': 2, '+G|+A|+A|+A|+G|=G': 1, '+A|+A|+G|=G': 1, '+C|+G|=G': 2, '+T|=G': 1, '+C|+A|=G': 1})\n",
      "=====================\n",
      "sample @ 1280\n",
      "defaultdict(<class 'int'>, {'=G': 819, '+A|=G': 64, '-G': 39, '+A|+G|=G': 16, 'N': 26, '+G|+A|=G': 1, '+T|+A|=G': 1, '+A|+T|+T|+C|+T|=G': 7, '+A|*GC': 1, '+C|+A|=G': 1, '+C|+A|+G|+A|=G': 1, '+A|+G|+A|=G': 1, '+G|=G': 2})\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from collections import Counter\n",
    "\n",
    "count = defaultdict(int)\n",
    "for cs in cssplits_control:\n",
    "    count[cs[i_ins]] += 1\n",
    "print(\"=====================\")\n",
    "print(f\"control @ {i_ins}\")\n",
    "print(count)\n",
    "\n",
    "count = defaultdict(int)\n",
    "for cs in cssplits_sample:\n",
    "    count[cs[i_ins]] += 1\n",
    "print(\"=====================\")\n",
    "print(f\"sample @ {i_ins}\")\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "left_idx, right_idx = set_indexes(sequence)\n",
    "count_5mer_sample = count_indels_5mer(cssplits_sample, left_idx, right_idx)\n",
    "count_5mer_control = count_indels_5mer(cssplits_control, left_idx, right_idx)\n",
    "sequence_errors = extract_sequence_errors(\n",
    "    count_5mer_sample, count_5mer_control, coverage_sample, coverage_control\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "65\n",
      "{'ins': [2, 5, 10, 5, 2], 'del': [9, 23, 130, 43, 6], 'sub': [1, 1, 1, 1, 1]}\n",
      "{'ins': [3, 6, 12, 4, 10], 'del': [2, 11, 126, 93, 8], 'sub': [1, 1, 1, 1, 1]}\n",
      "set()\n"
     ]
    }
   ],
   "source": [
    "print(i_del//5)\n",
    "print(count_5mer_sample[i_del//5])\n",
    "print(count_5mer_control[i_del//5])\n",
    "print(sequence_errors[i_del//5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9453921012935363 0.8592465758672474\n"
     ]
    }
   ],
   "source": [
    "samp = [c / coverage_sample for c in [9, 23, 130, 43, 6]]\n",
    "cont = [c / coverage_control for c in [2, 11, 126, 93, 8]]\n",
    "distance = 1 - cosine(samp, cont)\n",
    "_, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "print(distance, pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "256\n",
      "{'ins': [7, 7, 3, 9, 30], 'del': [1, 1, 1, 1, 1], 'sub': [1, 1, 1, 1, 1]}\n",
      "{'ins': [19, 15, 18, 23, 35], 'del': [1, 1, 1, 1, 1], 'sub': [1, 1, 1, 1, 1]}\n",
      "set()\n"
     ]
    }
   ],
   "source": [
    "print(i_ins//5)\n",
    "print(count_5mer_sample[i_ins//5])\n",
    "print(count_5mer_control[i_ins//5])\n",
    "print(sequence_errors[i_ins//5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9098501265616633 0.1041139955936684\n"
     ]
    }
   ],
   "source": [
    "samp = [c / coverage_sample for c in count_5mer_sample[i_ins//5][\"ins\"]]\n",
    "cont = [c / coverage_control for c in count_5mer_control[i_ins//5][\"ins\"]]\n",
    "distance = 1 - cosine(samp, cont)\n",
    "_, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "print(distance, pvalue)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 原因\n",
    "- 329も1280もdistanceが0.95以下であることによって補正を免れていた。\n",
    "    - → `extract_sequence_errors`における`distance`の閾値を0.95から0.9に変更した"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_sequence_errors(count_5mer_sample, count_5mer_control, coverage_sample, coverage_control):\n",
    "    sequence_errors = [set() for _ in range(len(count_5mer_sample))]\n",
    "    for i in range(len(sequence_errors)):\n",
    "        for mutation in [\"ins\", \"del\", \"sub\"]:\n",
    "            samp = [c / coverage_sample for c in count_5mer_sample[i][mutation]]\n",
    "            cont = [c / coverage_control for c in count_5mer_control[i][mutation]]\n",
    "            distance = 1 - cosine(samp, cont)\n",
    "            _, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "            if distance > 0.9 and pvalue > 0.05:\n",
    "                sequence_errors[i].add(mutation)\n",
    "    return sequence_errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_errors = extract_sequence_errors(\n",
    "    count_5mer_sample, count_5mer_control, coverage_sample, coverage_control\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'del'}\n"
     ]
    }
   ],
   "source": [
    "print(sequence_errors[i_del//5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ins'}\n"
     ]
    }
   ],
   "source": [
    "print(sequence_errors[i_ins//5])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## distance 0.9にすると、knokcinのright-loxpは検出できるのか？"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "\n",
    "import random\n",
    "import re\n",
    "from collections import Counter, defaultdict\n",
    "from copy import deepcopy\n",
    "from pathlib import Path\n",
    "import midsv\n",
    "from scipy import stats\n",
    "from scipy.spatial.distance import cosine\n",
    "\n",
    "\n",
    "def set_indexes(sequence: str):\n",
    "    sequence_length = len(sequence)\n",
    "    num_subset = sequence_length % 5\n",
    "    left_idx = 0\n",
    "    right_idx = sequence_length\n",
    "    if num_subset == 1:\n",
    "        left_idx += 1\n",
    "    elif num_subset == 2:\n",
    "        left_idx += 1\n",
    "        right_idx -= 1\n",
    "    elif num_subset == 3:\n",
    "        left_idx += 2\n",
    "        right_idx -= 1\n",
    "    elif num_subset == 4:\n",
    "        left_idx += 2\n",
    "        right_idx -= 2\n",
    "    return left_idx, right_idx\n",
    "\n",
    "\n",
    "def count_indels_5mer(cssplits: list[list[str]], left_idx: int, right_idx: int) -> list[dict]:\n",
    "    transposed = [list(t) for t in zip(*cssplits)]\n",
    "    count_indels_5mer = []\n",
    "    for i in range(left_idx, right_idx, 5):\n",
    "        count = {\"ins\": [1] * 5, \"del\": [1] * 5, \"sub\": [1] * 5}\n",
    "        cssplits_5mer = transposed[i : i + 5]\n",
    "        for j, cs in enumerate(cssplits_5mer):\n",
    "            counter = Counter(cs)\n",
    "            for key, cnt in counter.items():\n",
    "                if key.startswith(\"=\") or key == \"N\" or re.search(r\"a|c|g|t|n\", key):\n",
    "                    continue\n",
    "                if key.startswith(\"+\"):\n",
    "                    count[\"ins\"][j] += cnt\n",
    "                elif key.startswith(\"-\"):\n",
    "                    count[\"del\"][j] += cnt\n",
    "                elif key.startswith(\"*\"):\n",
    "                    count[\"sub\"][j] += cnt\n",
    "        count_indels_5mer.append(count)\n",
    "    return count_indels_5mer\n",
    "\n",
    "\n",
    "def extract_sequence_errors(count_5mer_sample, count_5mer_control, coverage_sample, coverage_control):\n",
    "    sequence_errors = [set() for _ in range(len(count_5mer_sample))]\n",
    "    for i in range(len(sequence_errors)):\n",
    "        for mutation in [\"ins\", \"del\", \"sub\"]:\n",
    "            samp = [c / coverage_sample for c in count_5mer_sample[i][mutation]]\n",
    "            cont = [c / coverage_control for c in count_5mer_control[i][mutation]]\n",
    "            distance = 1 - cosine(samp, cont)\n",
    "            _, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "            if distance > 0.9 and pvalue > 0.05:\n",
    "                sequence_errors[i].add(mutation)\n",
    "    return sequence_errors\n",
    "\n",
    "\n",
    "def replace_errors_to_atmark(cssplits_sample, sequence_errors, left_idx, right_idx):\n",
    "    cssplits_replaced = []\n",
    "    for samp in cssplits_sample:\n",
    "        samp_replaced = deepcopy(samp)\n",
    "        for idx_error, idx_5mer in enumerate(range(left_idx, right_idx, 5)):\n",
    "            samp_5mer = samp[idx_5mer : idx_5mer + 5]\n",
    "            error = sequence_errors[idx_error]\n",
    "            if \"ins\" in error:\n",
    "                samp_5mer = [\"@\" if cs.startswith(\"+\") else cs for cs in samp_5mer]\n",
    "            if \"del\" in error:\n",
    "                samp_5mer = [\"@\" if cs.startswith(\"-\") else cs for cs in samp_5mer]\n",
    "            if \"sub\" in error:\n",
    "                samp_5mer = [\"@\" if cs.startswith(\"*\") else cs for cs in samp_5mer]\n",
    "            samp_replaced[idx_5mer : idx_5mer + 5] = samp_5mer\n",
    "        cssplits_replaced.append(samp_replaced)\n",
    "    return cssplits_replaced\n",
    "\n",
    "\n",
    "def replace_atmark(cssplits: list[list[str]], sequence: str) -> list[list[str]]:\n",
    "    random.seed(1)\n",
    "    cssplits_replaced = deepcopy(cssplits)\n",
    "    sequence_length = len(sequence)\n",
    "    for i in range(1, sequence_length - 1):\n",
    "        cssplits_atmark = defaultdict(str)\n",
    "        cssplits_sampling_key = defaultdict(list)\n",
    "        cssplits_sampling_all = []\n",
    "        flag_all_atmark = True\n",
    "        for idx, cssplit in enumerate(cssplits):\n",
    "            key = \",\".join([cssplit[i - 1], cssplit[i + 1]])\n",
    "            if cssplit[i] == \"@\":\n",
    "                cssplits_atmark[idx] = key\n",
    "            else:\n",
    "                cssplits_sampling_key[key].append(cssplit[i])\n",
    "                cssplits_sampling_all.append(cssplit[i])\n",
    "                flag_all_atmark = False\n",
    "        for idx, key in cssplits_atmark.items():\n",
    "            if flag_all_atmark:\n",
    "                cssplits_replaced[idx][i] = \"N\"\n",
    "            elif cssplits_sampling_key[key]:\n",
    "                cssplits_replaced[idx][i] = random.choice(cssplits_sampling_key[key])\n",
    "            else:\n",
    "                cssplits_replaced[idx][i] = random.choice(cssplits_sampling_all)\n",
    "    for cs in cssplits_replaced:\n",
    "        if cs[0] == \"@\":\n",
    "            cs[0] = \"=\" + sequence[0]\n",
    "        if cs[-1] == \"@\":\n",
    "            cs[-1] = \"=\" + sequence[-1]\n",
    "    return cssplits_replaced\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "# main\n",
    "###############################################################################\n",
    "\n",
    "\n",
    "def execute(TEMPDIR: Path, FASTA_ALLELES: dict[str, str], CONTROL_NAME: str, SAMPLE_NAME: str) -> None:\n",
    "    for allele, sequence in FASTA_ALLELES.items():\n",
    "        midsv_sample = midsv.read_jsonl((Path(TEMPDIR, \"midsv\", f\"{SAMPLE_NAME}_splice_{allele}.jsonl\")))\n",
    "        midsv_control = midsv.read_jsonl((Path(TEMPDIR, \"midsv\", f\"{CONTROL_NAME}_splice_{allele}.jsonl\")))\n",
    "        coverage_sample = len(midsv_sample)\n",
    "        coverage_control = len(midsv_control)\n",
    "        cssplits_sample = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_sample]\n",
    "        cssplits_control = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_control]\n",
    "        left_idx, right_idx = set_indexes(sequence)\n",
    "        count_5mer_sample = count_indels_5mer(cssplits_sample, left_idx, right_idx)\n",
    "        count_5mer_control = count_indels_5mer(cssplits_control, left_idx, right_idx)\n",
    "        sequence_errors = extract_sequence_errors(\n",
    "            count_5mer_sample, count_5mer_control, coverage_sample, coverage_control\n",
    "        )\n",
    "        cssplits_sample_error_replaced = replace_errors_to_atmark(cssplits_sample, sequence_errors, left_idx, right_idx)\n",
    "        cssplits_control_error_replaced = replace_errors_to_atmark(\n",
    "            cssplits_control, sequence_errors, left_idx, right_idx\n",
    "        )\n",
    "        cssplits_sample_atmark_replaced = replace_atmark(cssplits_sample_error_replaced, sequence)\n",
    "        cssplits_control_atmark_replaced = replace_atmark(cssplits_control_error_replaced, sequence)\n",
    "        # Replace CSSPLIT\n",
    "        cssplits_sample_corrected = [\",\".join(cs) for cs in cssplits_sample_atmark_replaced]\n",
    "        cssplits_control_corrected = [\",\".join(cs) for cs in cssplits_control_atmark_replaced]\n",
    "        for i, cssplits in enumerate(cssplits_sample_corrected):\n",
    "            midsv_sample[i][\"CSSPLIT\"] = cssplits\n",
    "        for i, cssplits in enumerate(cssplits_control_corrected):\n",
    "            midsv_control[i][\"CSSPLIT\"] = cssplits\n",
    "        midsv.write_jsonl(midsv_control, Path(TEMPDIR, \"midsv\", f\"{CONTROL_NAME}_splice_{allele}.jsonl\"))\n",
    "        midsv.write_jsonl(midsv_sample, Path(TEMPDIR, \"midsv\", f\"{SAMPLE_NAME}_splice_{allele}.jsonl\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "execute(TEMPDIR, FASTA_ALLELES, CONTROL_NAME, SAMPLE_NAME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "classif_sample = classification.classify_alleles(TEMPDIR, SAMPLE_NAME)\n",
    "\n",
    "for classif in classif_sample:\n",
    "    classif[\"SV\"] = classification.detect_sv(classif[\"CSSPLIT\"], threshold=50)\n",
    "\n",
    "clust_sample = clustering.add_labels(classif_sample, TEMPDIR, CONTROL_NAME, FASTA_ALLELES, THREADS)\n",
    "clust_sample = clustering.add_readnum(clust_sample)\n",
    "clust_sample = clustering.add_percent(clust_sample)\n",
    "clust_sample = clustering.update_labels(clust_sample)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "both_loxp = \"f855fb9b57ee\"\n",
    "no_loxp = \"54661f075dc2\"\n",
    "left_loxp = \"7089ae32a686\"\n",
    "right_loxp = \"af9e73c7f917\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "both_loxp: 1\n",
      "right_loxp: 2\n",
      "no_loxp: 2\n",
      "left_loxp: 3\n"
     ]
    }
   ],
   "source": [
    "for samp in clust_sample:\n",
    "    if both_loxp in samp[\"QNAME\"]:\n",
    "        print(f\"both_loxp: {samp['LABEL']}\")\n",
    "    if left_loxp in samp[\"QNAME\"]:\n",
    "        print(f\"left_loxp: {samp['LABEL']}\")\n",
    "    if right_loxp in samp[\"QNAME\"]:\n",
    "        print(f\"right_loxp: {samp['LABEL']}\")\n",
    "    if no_loxp in samp[\"QNAME\"]:\n",
    "        print(f\"no_loxp: {samp['LABEL']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "from itertools import groupby\n",
    "from copy import deepcopy\n",
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "import midsv\n",
    "\n",
    "from src.DAJIN2.core.preprocess.correct_knockin import extract_knockin_loci\n",
    "from src.DAJIN2.core.clustering.preprocess import replace_both_ends_n, compress_insertion\n",
    "from src.DAJIN2.core.clustering.make_score import make_score\n",
    "from src.DAJIN2.core.clustering.annotate_score import annotate_score\n",
    "from src.DAJIN2.core.clustering.reorder_labels import reorder_labels\n",
    "from src.DAJIN2.core.clustering.return_labels import return_labels\n",
    "\n",
    "\n",
    "paths_midsv = list(Path(TEMPDIR, \"midsv\").glob(f\"{CONTROL_NAME}_splice_*\"))\n",
    "cssplits_control_by_alleles = defaultdict(list)\n",
    "for path_midsv in paths_midsv:\n",
    "    midsv_control = midsv.read_jsonl(path_midsv)\n",
    "    allele = path_midsv.stem.split(\"_\")[-1]\n",
    "    cssplits = [cs[\"CSSPLIT\"].split(\",\") for cs in midsv_control]\n",
    "    cssplits_control_by_alleles[allele] = cssplits\n",
    "knockin_alleles = extract_knockin_loci(TEMPDIR)\n",
    "labels_all = []\n",
    "max_label = 0\n",
    "\n",
    "allele = \"control\"\n",
    "sv = False\n",
    "\n",
    "classif_sample.sort(key=lambda x: (x[\"ALLELE\"], x[\"SV\"]))\n",
    "cssplits_control = cssplits_control_by_alleles[allele]\n",
    "cssplits_sample = [cs[\"CSSPLIT\"].split(\",\") for cs in classif_sample if cs[\"ALLELE\"] == allele and cs[\"SV\"] == sv]\n",
    "cssplits_control = replace_both_ends_n(cssplits_control)\n",
    "cssplits_sample = replace_both_ends_n(cssplits_sample)\n",
    "cssplits_control = compress_insertion(cssplits_control)\n",
    "cssplits_sample = compress_insertion(cssplits_sample)\n",
    "mutation_score = make_score(cssplits_control, cssplits_sample, knockin_alleles[allele])\n",
    "scores_control = annotate_score(cssplits_control, mutation_score)\n",
    "scores_sample = annotate_score(cssplits_sample, mutation_score)\n",
    "labels = return_labels(scores_sample, scores_control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({1: 372})"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f544af5ce50>]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAD4CAYAAADFAawfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAOHklEQVR4nO3df6xkdXnH8c9n9y4oC6i4I9kA6wVjTDYmCkywhIaktEGgTW2jf2Diz5jcf9RgqjEQosH0D2OT2ta0IblVWrRU0ogEQ1AhKiH+QmdhwV0XZKEYVpAdYpQFZZfdffrHnKuH23PvnDvOmXlmzvuVTO7cM2eG5zuH/ex3nvOdPY4IAQDy2jTtAgAA6yOoASA5ghoAkiOoASA5ghoAklto4kW3bdsWi4uLTbw0AMylXbt2PRMRnarHGgnqxcVF9Xq9Jl4aAOaS7Z+v9RitDwBIjqAGgOQIagBIjqAGgOQIagBIjqAGgOQIagBIjqAGEnn+8FHdev+BaZeBZBr5wguA0Xzytr265b4D2nHaVp3/2ldNuxwkwYwaSOTpZ1+QJP32yNEpV4JMCGoASI6gBoDkCGoASI6gBoDkCGoASI6gBoDkCGoASI6gBoDkCGoASI6gBoDkCGoASI6gBhKKmHYFyISgBoDkCGogIXvaFSCTWkFt+5W2v2L7Idv7bF/YdGEAgIG6Fw74F0nfiIh32D5B0kkN1gQAKBka1LZPlXSxpPdJUkQckXSk2bIAACvqtD7OkdSX9B+277f9edtbV+9ke8l2z3av3++PvVAAaKs6Qb0g6TxJ10fEuZKel3T16p0iYjkiuhHR7XQ6Yy4TaBeW56GsTlAfkHQgIu4tfv+KBsENAJiAoUEdEb+U9ITtNxSb/lzSTxutCmg5luehrO6qjw9LuqlY8fGYpPc3VxIAoKxWUEfEbkndZksBAFThm4kAkBxBDQDJEdRAQizPQxlBDQDJEdRAQizPQxlBDQDJEdQAkBxBDQDJEdQAkBxBDSTE8jyUEdQAkBxBDSTE8jyUEdQAkBxBDQDJEdQAkBxBDSTEqg+UEdQAkBxBDQDJEdRAQizPQxlBDQDJ1bq4re3HJR2SdEzS0YjgQrcAMCG1grrwZxHxTGOVAAAq0foAEmJ5HsrqBnVIutP2LttLTRYEtBknEVGlbuvjooh40vZrJN1l+6GIuKe8QxHgS5K0Y8eOMZcJtAMzaVSpNaOOiCeLnwcl3Srpgop9liOiGxHdTqcz3iqBlmFmjbKhQW17q+1TVu5LulTSnqYLAwAM1Gl9nC7pVg/+il+Q9N8R8Y1GqwIA/N7QoI6IxyS9aQK1AAAqsDwPSIiTiigjqIFEOImIKgQ1kAgzaVQhqIGEmFmjjKAGgOQIagBIjqAGgOQIaiAhTiqijKAGEuEkIqoQ1EAizKRRhaAGEmJmjTKCGgCSI6gBIDmCGgCSI6iBhDipiDKCGkiEk4ioQlADiTCTRhWCGkiImTXKCGoASI6gBoDkCGogIXrVKKsd1LY3277f9u1NFgQAeKmNzKivkrSvqUIAcBIR1WoFte0zJf2lpM83Ww7QbrQ8UKXujPqfJX1c0vG1drC9ZLtnu9fv98dRG9BazKxRNjSobf+VpIMRsWu9/SJiOSK6EdHtdDpjKxAA2q7OjPoiSX9t+3FJN0u6xPZ/NVoVAOD3hgZ1RFwTEWdGxKKkKyV9OyLe1XhlQIvRq0YZ66gBILmFjewcEXdLuruRSgBwEhGVmFEDidDyQBWCGkiImTXKCGoASI6gBoDkCGogIXrVKCOogUToTaMKQQ0AyRHUQCK0PFCFoAaA5AhqICF61SgjqAEgOYIaSIheNcoIaiARWh6oQlADQHIENZAILQ9UIagBIDmCGkiIXjXKCGoASI6gBhKiV40yghpIhJYHqhDUAJDc0KC2/TLbP7L9gO29tj81icKANqLlgSoLNfY5LOmSiHjO9hZJ37X99Yj4YcO1AQBUI6gjIiQ9V/y6pbjx9z7QIHrVKKvVo7a92fZuSQcl3RUR91bss2S7Z7vX7/fHXCbQLrRAUFYrqCPiWES8WdKZki6w/caKfZYjohsR3U6nM+YyAaC9NrTqIyJ+LeluSZc1UQzQdrQ8UKXOqo+O7VcW918u6S8kPdRwXQCAQp1VH9sl3Wh7swbB/j8RcXuzZQHtRG8aVeqs+nhQ0rkTqAUAUIFvJgIJ0atGGUENJDS0BXLnJ6TrXjGRWjB9BDUwi77/uWlXgAkiqIFEaHmgCkENAMkR1EAiLM9DFYIaAJIjqIGE6FWjjKAGEqIFgjKCGkiEmTSqENQAkBxBDQDJEdRAIvSmUYWgBoDkCGogIU4qooygBhKiBYIyghpIhJk0qhDUAJAcQQ0AyRHUQCL0plGFoAaA5IYGte2zbH/H9j7be21fNYnCgDbjpCLKFmrsc1TSRyPiPtunSNpl+66I+GnDtQGtRQsEZUNn1BHxVETcV9w/JGmfpDOaLgxoI2bSqLKhHrXtRUnnSrq34rEl2z3bvX6/P6byAKyLqXcr1A5q2ydLukXSRyLi2dWPR8RyRHQjotvpdMZZIwC0Wq2gtr1Fg5C+KSK+2mxJQHtteILMjLoV6qz6sKQvSNoXEZ9tviQAQFmdGfVFkt4t6RLbu4vbFQ3XBQAoDF2eFxHflcS5aCAlWh9twDcTgURYnocqBDUAJEdQA7OMVR+tQFADiZC7qEJQA0ByBDUw05iCtwFBDQDJEdRAIizPQxWCGphlnH1sBYIaAJIjqIFEmCCjCkENzDSSvQ0IagBIjqAGZhm9klYgqIFEWJ6HKgQ1ACRHUAMzjdZHGxDUAJAcQQ0kwrlBVCGogVlGsrcCQQ0kwqoPVBka1LZvsH3Q9p5JFIT8fvO7F/Wzpw9NuwygNerMqP9T0mUN14EZ8vbrv69L/+meaZcBSaz6aIehQR0R90j61QRqwYzYf/C5aZcAtMrYetS2l2z3bPf6/f64XhYAWm9sQR0RyxHRjYhup9MZ18sCrbLhRRys+mgFVn0AQHIENZDIxpfnMaNugzrL874s6QeS3mD7gO0PNF8WAGDFwrAdIuKdkygEAFCN1gcwyziZ2AoENQAkR1ADiTBBRhWCGphpJHsbENRAIvzreahCUANAcgQ1RhY0VKePY9AKBDVGRkYAk0FQY2TkNDAZBDVGRutj/Db+lnIM2oCgBoDkCGqMjLnc+G14eR6falqBoMbIyAhgMghqjCyYUwMTQVBjZMyogckgqAEgOYIaI2NGPX68p6hCUGNk9KgTINlbgaDGyMiI8eNfz0MVghojI6eBySCoMTK+Qp4Bx6ANagW17ctsP2x7v+2rmy4KAPAHQ4Pa9mZJ/ybpckk7Jb3T9s6mC0N+zOWAyfCwj6+2L5R0XUS8tfj9GkmKiE+v9Zxutxu9Xm/DxTzy9+drSxze8PMwWUeOHpcknbBA52zcVt7bhU3Wpk1rn1lcPP6EJOmJTWfoGB3MNH67+RXaee33Rnqu7V0R0a16bKHG88+Q9ETp9wOS3lLxH1mStCRJO3bsGKFM6TdbF7Xp+JGRnovJef7wMT37wovaftLLpl3K3Dl6LNR/7rC2n7z+e7vpxRO14/B+9U963YQqQx1Ht5zayOvWCeqqv9b/3zQ8IpYlLUuDGfUoxXT/7pZRnga01nnTLgATUecz0wFJZ5V+P1PSk82UAwBYrU5Q/1jS622fbfsESVdK+lqzZQEAVgxtfUTEUdsfkvRNSZsl3RARexuvDAAgqV6PWhFxh6Q7Gq4FAFCBdT0AkBxBDQDJEdQAkBxBDQDJDf0K+Ugvavcl/XzEp2+T9MwYy8mG8c2ueR6bxPim7bUR0al6oJGg/mPY7q31ffd5wPhm1zyPTWJ8mdH6AIDkCGoASC5jUC9Pu4CGMb7ZNc9jkxhfWul61ACAl8o4owYAlBDUAJBcmqCelwvo2n7c9k9s77bdK7adZvsu248UP19V2v+aYswP237r9CqvZvsG2wdt7ylt2/B4bJ9fvC/7bX/O9trXmZqgNcZ3ne1fFMdwt+0rSo/NzPhsn2X7O7b32d5r+6pi+1wcv3XGNxfH7yUiYuo3Df751EclnSPpBEkPSNo57bpGHMvjkrat2vYPkq4u7l8t6TPF/Z3FWE+UdHbxHmye9hhW1X6xBhcS2fPHjEfSjyRdqMEVg74u6fJpj22d8V0n6WMV+87U+CRtl3Recf8UST8rxjAXx2+d8c3F8SvfssyoL5C0PyIei4gjkm6W9LYp1zROb5N0Y3H/Rkl/U9p+c0Qcjoj/lbRfg/cijYi4R9KvVm3e0Hhsb5d0akT8IAZ/Kr5Yes5UrTG+tczU+CLiqYi4r7h/SNI+Da6BOhfHb53xrWWmxleWJairLqC73hueWUi60/au4oK/knR6RDwlDf7nkvSaYvusjnuj4zmjuL96e2Yfsv1g0RpZaQ3M7PhsL0o6V9K9msPjt2p80pwdvyxBXesCujPioog4T9Llkj5o++J19p2ncUtrj2fWxnm9pNdJerOkpyT9Y7F9Jsdn+2RJt0j6SEQ8u96uFdtmcXxzdfykPEE9NxfQjYgni58HJd2qQSvj6eLjlYqfB4vdZ3XcGx3PgeL+6u0pRcTTEXEsIo5L+nf9oR01c+OzvUWDELspIr5abJ6b41c1vnk6fiuyBPVcXEDX9lbbp6zcl3SppD0ajOW9xW7vlXRbcf9rkq60faLtsyW9XoOTGtltaDzFx+tDtv+kOJv+ntJz0lkJscLfanAMpRkbX1HLFyTti4jPlh6ai+O31vjm5fi9xLTPZpbOyF6hwVnbRyVdO+16RhzDORqcVX5A0t6VcUh6taRvSXqk+Hla6TnXFmN+WMnONBf1fVmDj48vajDz+MAo45HU1eAPzKOS/lXFt2KnfVtjfF+S9BNJD2rwh3v7LI5P0p9q8BH+QUm7i9sV83L81hnfXBy/8o2vkANAcllaHwCANRDUAJAcQQ0AyRHUAJAcQQ0AyRHUAJAcQQ0Ayf0fycuw15KcrrcAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.plot(scores_sample[74])\n",
    "plt.plot(scores_sample[118])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.exceptions import ConvergenceWarning\n",
    "from collections import Counter\n",
    "\n",
    "from src.DAJIN2.core.clustering.merge_clusters import merge_clusters\n",
    "from src.DAJIN2.core.clustering.reorder_labels import reorder_labels\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=ConvergenceWarning)\n",
    "\n",
    "###############################################################################\n",
    "# Dimension reduction\n",
    "###############################################################################\n",
    "\n",
    "\n",
    "def reduce_dimension(scores_sample: list[list], scores_control_subset: list[list]) -> np.array:\n",
    "    scores = scores_sample + scores_control_subset\n",
    "    n_components = min(20, len(scores[0]))\n",
    "    pca = PCA(n_components=n_components).fit(scores)\n",
    "    return pca.transform(scores)\n",
    "\n",
    "\n",
    "def optimize_labels(X: np.array, scores_sample: list[list], scores_control_subset: list[list]) -> list[int]:\n",
    "    scores = scores_sample + scores_control_subset\n",
    "    n_components = min(20, len(scores))\n",
    "    labels_results = [1] * len(scores_sample)\n",
    "    for i in range(1, n_components):\n",
    "        np.random.seed(seed=1)\n",
    "        labels = GaussianMixture(n_components=i, random_state=1).fit_predict(X)\n",
    "        labels = labels.tolist()\n",
    "        labels_sample = labels[: len(scores_sample)]\n",
    "        labels_control = labels[len(scores_sample) :]\n",
    "        labels_merged = merge_clusters(labels_control, labels_sample)\n",
    "        labels_reorder = reorder_labels(labels_merged)\n",
    "        # Reads < 1% in the control are considered clustering errors and are not counted\n",
    "        count_control = Counter(labels_control)\n",
    "        num_labels_control = sum(1 for reads in count_control.values() if reads/sum(count_control.values())*100 > 1)\n",
    "        if num_labels_control > 1:\n",
    "            return labels_results\n",
    "        labels_results = labels_reorder\n",
    "    return labels_results\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "# main\n",
    "###############################################################################\n",
    "\n",
    "\n",
    "def return_labels(scores_sample: list[list], scores_control: list[list]) -> list[int]:\n",
    "    np.random.seed(seed=1)\n",
    "    X_control = reduce_dimension([], scores_control)\n",
    "    labels = GaussianMixture(n_components=2, random_state=1).fit_predict(X_control)\n",
    "    label_most = Counter(labels).most_common()[0][0]\n",
    "    scores_control_subset = [s for l, s in zip(labels, scores_control) if l == label_most][:1000]\n",
    "    X = reduce_dimension(scores_sample, scores_control_subset)\n",
    "    labels = optimize_labels(X, scores_sample, scores_control_subset)\n",
    "    return labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(seed=1)\n",
    "X_control = reduce_dimension([], scores_control)\n",
    "labels = GaussianMixture(n_components=2, random_state=1).fit_predict(X_control)\n",
    "label_most = Counter(labels).most_common()[0][0]\n",
    "scores_control_subset = [s for l, s in zip(labels, scores_control) if l == label_most][:1000]\n",
    "X = reduce_dimension(scores_sample, scores_control_subset)\n",
    "labels = optimize_labels(X, scores_sample, scores_control_subset)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 Counter({0: 372}) Counter({0: 912})\n",
      "2 Counter({1: 192, 0: 180}) Counter({0: 712, 1: 200})\n",
      "3 Counter({1: 177, 0: 165, 2: 30}) Counter({0: 692, 1: 208, 2: 12})\n",
      "4 Counter({1: 177, 0: 165, 3: 26, 2: 4}) Counter({0: 692, 1: 208, 3: 12})\n",
      "5 Counter({0: 154, 1: 153, 4: 35, 3: 26, 2: 4}) Counter({0: 701, 1: 199, 3: 12})\n",
      "6 Counter({0: 153, 1: 137, 4: 35, 3: 26, 5: 17, 2: 4}) Counter({0: 700, 1: 184, 5: 16, 3: 12})\n",
      "7 Counter({0: 148, 1: 124, 4: 34, 3: 26, 6: 19, 5: 17, 2: 4}) Counter({0: 690, 1: 189, 5: 16, 3: 12, 6: 5})\n",
      "8 Counter({1: 112, 0: 84, 6: 72, 4: 34, 2: 30, 3: 17, 5: 12, 7: 11}) Counter({1: 629, 0: 146, 6: 99, 3: 16, 2: 12, 7: 10})\n",
      "9 Counter({1: 105, 0: 79, 6: 75, 4: 34, 2: 30, 3: 17, 5: 12, 7: 11, 8: 9}) Counter({1: 614, 0: 137, 6: 107, 8: 16, 3: 16, 2: 12, 7: 10})\n",
      "10 Counter({1: 105, 0: 79, 6: 75, 4: 34, 2: 30, 9: 16, 5: 12, 7: 11, 8: 9, 3: 1}) Counter({1: 614, 0: 137, 6: 107, 8: 16, 9: 16, 2: 12, 7: 10})\n",
      "11 Counter({1: 164, 0: 78, 4: 34, 2: 30, 9: 15, 5: 12, 6: 12, 7: 11, 8: 9, 10: 6, 3: 1}) Counter({1: 702, 0: 134, 6: 18, 9: 16, 8: 15, 2: 12, 7: 10, 10: 5})\n",
      "12 Counter({1: 191, 0: 40, 4: 34, 2: 30, 9: 15, 5: 12, 6: 12, 11: 11, 7: 11, 8: 9, 10: 6, 3: 1}) Counter({1: 760, 0: 64, 6: 18, 9: 16, 8: 15, 11: 12, 2: 12, 7: 10, 10: 5})\n",
      "13 Counter({1: 180, 0: 45, 4: 34, 2: 30, 9: 15, 5: 12, 6: 12, 11: 11, 7: 11, 8: 9, 10: 6, 12: 6, 3: 1}) Counter({1: 755, 0: 64, 6: 18, 9: 16, 8: 15, 11: 12, 2: 12, 7: 10, 10: 5, 12: 5})\n",
      "14 Counter({1: 180, 0: 45, 4: 34, 2: 26, 9: 15, 5: 12, 6: 12, 11: 11, 7: 11, 8: 9, 10: 6, 12: 6, 13: 4, 3: 1}) Counter({1: 755, 0: 64, 6: 18, 9: 16, 8: 15, 11: 12, 2: 12, 7: 10, 10: 5, 12: 5})\n",
      "15 Counter({1: 180, 0: 45, 4: 34, 2: 26, 9: 15, 14: 12, 6: 12, 11: 11, 7: 10, 8: 9, 10: 6, 12: 6, 13: 4, 3: 1, 5: 1}) Counter({1: 755, 0: 64, 6: 18, 9: 16, 8: 15, 11: 12, 2: 12, 7: 10, 10: 5, 12: 5})\n",
      "16 Counter({1: 158, 0: 59, 4: 34, 2: 26, 9: 15, 14: 12, 6: 12, 11: 11, 7: 10, 8: 9, 15: 8, 10: 6, 12: 6, 13: 4, 3: 1, 5: 1}) Counter({1: 699, 0: 107, 6: 18, 9: 16, 8: 15, 15: 13, 11: 12, 2: 12, 7: 10, 10: 5, 12: 5})\n",
      "17 Counter({1: 144, 0: 69, 4: 34, 2: 26, 9: 15, 14: 12, 11: 11, 6: 10, 7: 10, 8: 9, 15: 8, 10: 6, 16: 6, 12: 6, 13: 4, 3: 1, 5: 1}) Counter({1: 674, 0: 127, 6: 18, 9: 16, 8: 15, 15: 13, 11: 12, 2: 12, 7: 10, 10: 5, 16: 5, 12: 5})\n",
      "18 Counter({1: 120, 0: 87, 4: 34, 2: 26, 9: 15, 14: 12, 11: 10, 7: 10, 6: 9, 17: 9, 8: 9, 15: 8, 10: 6, 16: 6, 12: 5, 13: 4, 3: 1, 5: 1}) Counter({1: 632, 0: 162, 6: 18, 9: 16, 8: 15, 11: 12, 15: 12, 2: 12, 7: 10, 17: 8, 10: 5, 16: 5, 12: 5})\n",
      "19 Counter({1: 120, 0: 87, 4: 34, 2: 26, 9: 15, 14: 12, 11: 10, 7: 10, 6: 9, 17: 9, 8: 9, 15: 8, 10: 6, 16: 6, 12: 5, 18: 3, 13: 1, 3: 1, 5: 1}) Counter({1: 632, 0: 162, 6: 18, 9: 16, 8: 15, 11: 12, 15: 12, 2: 12, 7: 10, 17: 8, 10: 5, 16: 5, 12: 5})\n"
     ]
    }
   ],
   "source": [
    "scores = scores_sample + scores_control_subset\n",
    "n_components = min(20, len(scores))\n",
    "labels_results = [1] * len(scores_sample)\n",
    "for i in range(1, n_components):\n",
    "    np.random.seed(seed=1)\n",
    "    labels = GaussianMixture(n_components=i, random_state=1).fit_predict(X)\n",
    "    labels = labels.tolist()\n",
    "    labels_sample = labels[: len(scores_sample)]\n",
    "    labels_control = labels[len(scores_sample) :]\n",
    "    labels_merged = merge_clusters(labels_control, labels_sample)\n",
    "    labels_reorder = reorder_labels(labels_merged)\n",
    "    print(i, Counter(labels_sample), Counter(labels_control)) # ! ========================\n",
    "    # # Reads < 1% in the control are considered clustering errors and are not counted\n",
    "    # count_control = Counter(labels_control)\n",
    "    # num_labels_control = sum(1 for reads in count_control.values() if reads/sum(count_control.values())*100 > 1)\n",
    "    # if num_labels_control > 1:\n",
    "    #     return labels_results\n",
    "    # labels_results = labels_reorder\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 Counter({1: 192, 0: 180}) Counter({0: 712, 1: 200})\n"
     ]
    }
   ],
   "source": [
    "i = 2\n",
    "np.random.seed(seed=1)\n",
    "labels = GaussianMixture(n_components=i, random_state=1).fit_predict(X)\n",
    "labels = labels.tolist()\n",
    "labels_sample = labels[: len(scores_sample)]\n",
    "labels_control = labels[len(scores_sample) :]\n",
    "labels_merged = merge_clusters(labels_control, labels_sample)\n",
    "labels_reorder = reorder_labels(labels_merged)\n",
    "print(i, Counter(labels_sample), Counter(labels_control)) # ! ========================\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### まだ補正が足りない？？？\n",
    "controlとright-loxpの２群に分かれてほしいのに…"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 1, 0, 1, 1, 1, 0, 1, 0, 1]\n"
     ]
    }
   ],
   "source": [
    "print(labels_sample[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set()\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_formats = {'png', 'retina'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f5449537820>]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtkAAAH2CAYAAAClX5DgAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAABYlAAAWJQFJUiTwAAAlgUlEQVR4nO3dfZhkV10n8G9nOp1MmDABGV4DGhAO6IryDjFsQEECLqsCoqsgQTGr7uq6sAvoIy4i+qA8ymKWrIKKy4KruDygixBE2RAiIBAlAoHDIgRJkDAEM2QmyfS89P5xq8hM09N9e+pUV1fV5/M8eW6q6vY9t/o3t+tb55577sLKykoAAIB2Tpn0DgAAwKwRsgEAoDEhGwAAGhOyAQCgMSEbAAAaE7IBAKAxIRsAABoTsgEAoDEhGwAAGhOyAQCgMSEbAAAaE7IBAKCxxUnvwEn6uyTnJNmf5FMT3hcAAGbTNybZleQzSR60mR9cWFlZGcsejdmNSXZPeicAAJgL+5KctZkfmNae7P1Jdh89upLDh49sWaNLS92va3n58Ja1STvqN93Ub3qp3XRTv+mmfqNZXNyRU05ZSLrsubmfbb87W+JTSe5x+PCR7Nt3y5Y1umfPmUmypW3SjvpNN/WbXmo33dRvuqnfaHbv3jn8orLp4ckufAQAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBoBNWllZmfQuANvc4qR3AACmxcrKSl77sT/MP+y7Jj90/6fmm7/u/pPeJWCb0pMNAD199IaP58ovXpUbD+7LJVf9/qR3B9jGhGwA6OkLB7446V0ApoSQDQAAjQnZAADQmJANAACNCdkAANCYkA0AAI0J2QAA0JiQDQAAjQnZAADQmJANAACNCdkAANCYkA0AAI0J2QAA0JiQDQAAjQnZAADQmJANAACNCdkAANCYkA0AAI0J2QAA0JiQDQAAjQnZAADQmJANAACNCdkAANCYkA0AAI0J2QAA0JiQDQAAjQnZAADQmJANAACNCdkAANDYYqsNlVK+PskvJnlCkjsn2Zvkz5P8Yq31C63aAQCA7a5JT3Yp5aFJrkryo0m+nC5cH03y40muKKXcoUU7AAAwDUYO2aWU05L8YZLdSX6m1vrAWuv3JblvkjcluU+SF4/aDgAATIsWPdlPTxeo31BrvXj4ZK311iT/Mcn1SUqDdgAAYCq0GJP91MHyN1e/UGv9XJK7NmgDAACmRouQ/eAky0muKqXcM8kPJfnGJDckeVOt9YMN2gAAgKkxUsgejMe+Z5Jrk3x/kt9LcsYxq7yglPLyWuvzR2nnRJaWFrNnz5nj2PS6JtEm7ajfdFO/6TULtdt1w2nHPZ6F99TXPL3XWaR+W2/UMdm3HyzvmOR1Sd6cbvz1HZL8YLqZRv5zKeWiEdsBAICpMepwkdMHyzOS/EWt9RnHvPbHpZT9Sd6a5BdLKa+pta6M2N5xlpcPZ9++W1pucl3Db4F79960ZW3SjvpNN/WbXrNUu/37Dx73eBbe00ZmqX7zSP1Gs3v3ziwtnVxcHrUn+8Ax/3/J6hdrrX+e5Lok90g3ThsAAGbeqCF7X7qLHpPkmhOs89nB8k4jtgUAAFNhpJBdaz2S5OODh3c/wWrDKfz2jtIWAABMixY3o3n7YPn01S+UUkqSb0jy+SSfbtAWAABsey1C9m+nG5v9I6WUHxo+WUq5Q5LfHbTxqlrr0QZtAQDAtjdyyK61fjbJjyY5muQNpZQrSyl/luSTSc5L8q4kLx+1HQAAmBYterJTa31jkocleVOSeyV5fJIvJnlhkgtqrYdatAMAANOgxW3VkyS11g8neVqr7QEAwLRq0pMNAADcRsgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABobLHFRkopz0zyunVW+ZVa6y+0aAsAALa7JiE7yYMGy3cm+eIar3+4UTsAALDttQ7Zz661XtdomwAAMJVajcn+tiTXC9gAANAgZJdSzklyVpIrR94bAACYAS2GiwyHilxfSrk4yROTnJ3ks0len+TltdZbG7QDAABToWXIfnaSf07yniTXJXlokpckuaCU8rha6y0N2jrO0tJi9uw5s/VmNzSJNmlH/aab+k2vWajdrhtOP+7xLLynvubpvc4i9dt6LcZkD0P2G5Pcs9b6PbXW85N8c5Krkpyb5KUN2gGACVuZ9A4AU6JFT/bTktw7yadqrcvDJ2ut15RSLkzyt0kuKqW8sNZ6qEF7X7W8fDj79jXvID+h4bfAvXtv2rI2aUf9ppv6Ta9Zqt3+/QePezwL72kjs1S/eaR+o9m9e2eWlk4uLo8csgfjra8+wWsfLqVcm+SeSe6X5GOjtgcAANvdVtxW/QuD5Rlb0BYAAEzcSD3ZpZQzk/xGkjsm+cFa6+E1VjtnsDSHNgAAc2HUnuz9Sb4vyVOTnL/6xVLKBUnulOQjtdbPj9gWAABMhZFCdq11JclrBg8vLqXcffhaKeU+SS4ZPDS7CABTb2FhYdK7AEyJFrOL/HKSRyc5L0ktpVwxeP6xSU5L8pu11jc2aAcAJmplxRR+QD8jX/g4uMnMdyZ5YZJr0oXrc5O8P8lTa63PG7UNAACYJi16sjOYH/vXBv8BAMBc24op/AAAYK4I2QAA0JiQDQAAjQnZANCTKfyAvoRsAOjJFH5AX0I2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2APS0sLAw6V0ApoSQDQA9raysTHoXgCkhZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBjQjYA9GQKP6AvIRsAejKFH9CXkA0AAI0J2QAA0JiQDQAAjQnZAADQmJANAACNCdkA0JMp/IC+hGwA6MkUfkBfQjYAADQmZAMAQGNCNgAANCZkAwBAY4vj2Ggp5Y5JPprkbrVWl2IDADBXxtWTfUmSu41p2wAwEabwA/pqHrJLKf8myQ+03i4ATJop/IC+mobsUsrdk/y3JO9NcqTltgEAYFq07sn+vSSnJ3lW4+0CAMDUaBaySyk/meSCJC+otX6q1XYBAGDaNJldpJRynyQvT/KuJK9qsc0+lpYWs2fPmVvV3FdNok3aUb/ppn7TaxZqt+uG0457PAvvqa95eq+zSP223sg92aWUHUlel+RokmfXWl0VAgDAXGvRk/38JOcmeU6t9R8bbK+35eXD2bfvli1rb/gtcO/em7asTdpRv+mmftNrlmp34MDycY9n4T1tZJbqN4/UbzS7d+/M0tLJxeWRerJLKd+a5MVJ3lZr/b1RtgUA250p/IC+Ru3J/pUkS0lOLaW8ftVrpyTJMc//bK31SyO2BwAA296oIXvXYPn4ddb54cHyF5II2QAAzLyRQnat9TEneq2UcjjJjlqre9ACADBXmt9WHQAA5p2QDQAAjQnZANDTwoIRkEA/Te74uJZa69i2DQCTYAo/oC892QAA0JiQDQAAjQnZAADQmJANAACNCdkA0JPZRYC+hGwAAGhMyAaAnkzhB/QlZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDQE+m8AP6ErIBAKAxIRsAejKFH9CXkA0AAI0J2QAA0JiQDQAAjQnZAADQmJANAD2Zwg/oS8gGAIDGhGwA6MkUfkBfQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2APRkCj+gLyEbAHoyuwjQl5ANAACNCdkAANCYkA0AAI0J2QAA0JiQDQAAjQnZANCTKfyAvoRsAOjJFH5AX0I2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgD0ZHYRoC8hGwAAGhOyAaAnU/gBfQnZAADQ2GKLjZRSdiT5d0l+LElJcnOSDyV5Za31z1u0AQAA06JVT/Zrk7wyyTck+askVyY5P8lbSykvatQGAABMhZFDdinl6UmemaQmuV+t9btrrY9P8uAk+5K8uJRy31HbAQCAadGiJ/sZg+ULa63XD5+stX4syRsGbXxXg3YAYKJM4Qf01WJM9tOS3C/J/1vjtTMHy8MN2gEAgKkwcsiutS4n+ejq50sp/yrJ9yfZn+Qto7YDAJNmCj+gryaziwyVUnYm+Z9JvinJA5L8Y5JnHjuMBAAAZl3TkJ3kXkmeuuq5Bya5vHE7SZKlpcXs2XPmxis2Nok2aUf9ppv6Ta9ZqN2uG0477vEsvKe+5um9ziL123qtb0ZzbZI7JbljkqcnOTXJxaWUFzRuBwAAtq2mPdm11gNJDgwe/kkp5XNJ3pvk50spr6y13tqyveXlw9m375aWm1zX8Fvg3r03bVmbtKN+0039ptcs1W7//oPHPZ6F97SRWarfPFK/0ezevTNLSycXl8d6W/Va6/uT/EOS2ye59zjbAoBxM4Uf0NdIPdmllIUkv5ZuLPYzaq1rTdU3/Np/6ihtAQDAtBipJ7vWupLke5P8QNa44Uwp5ZwkJd0QkjpKWwAwaabwA/pqMVzk1YPlb5VSzh4+WUq5R5I/Stdb/qrW47EBAGC7anHh4yuTPDbJk5J8opRyxWC7j0iyK8nbkryoQTsAADAVRu7JrrUeSvKvk/xMkk8mOT/Jo5J8LMlPJHny4K6QAAAwF5pM4VdrPZLk4sF/AAAw18Y6hR8AzBJT+AF9CdkAANCYkA0APZnCD+hLyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGgJ5M4Qf0JWQDAEBjQjYA9GQKP6AvIRsAABoTsgEAoDEhGwAAGhOyAQCgMSEbAHoyhR/Ql5ANAACNCdkA0JMp/IC+hGwAAGhMyAYAgMaEbAAAaEzIBoCezC4C9CVkAwBAY0I2AAA0JmQDQE+m8AP6ErIBAKAxIRsAABoTsgEAoDEhGwB6MoUf0JeQDQAAjQnZAMDIDh85OuldgG1FyAaAnkzht7bf+bOP5d//18tz+VWfn/SuwLYhZAMAJ+0z//SV/M3V12f50NH8wds/MendgW1DyAYATtq+/cuT3gXYloRsAABoTMgGgJ5M4Qf0JWQDAEBjQjYAADQmZANAT6bwA/oSsgEAoDEhGwAAGhOyAQCgMSEbAHoyhR/Ql5ANAACNLbbYSCllR5KfTPKsJA9IsiPJp5P8UZKX11pvbdEOAEyS2UWAvkbuyR4E7D9NcnGS+yd5f5LLktw9yUuSXFZKOWPUdgAAYFq0GC7ynCTfneTvk9y/1vq4WusTk9w3yfuSPCLJixq0AwAAU6FFyL5wsPzZWut1wydrrV9KN4QkSX6wQTsAADAVWoTsLyX5RJIPrPHaJwfLuzdoBwAApsLIFz7WWp+8zssPGyyvHbUdAJg0U/gBfTWZXWQtpZSFdBc+JsmbxtHG0tJi9uw5cxybXtck2qQd9Ztu6je9ZqF2t7vhtOMez8J76utE73X3Fw/0Wo/JUpetN855sn81yflJrk/y8jG2AwBbxBR+QD9j6ckupbwkyQuTHEzy9Frr3nG0s7x8OPv23TKOTa9p+C1w796btqxN2lG/6aZ+02uWard//8HjHs/Ce9rIRvVb/Tk8D7+TaTJLx98k7N69M0tLJxeXm4bsUspiklcluSjJrUmeUmu9vGUbAACw3TUL2aWUXUn+JMkFSW5M8j0CNgAA86jVbdXvkOSdSR6S5HNJnlRr/WiLbQPAdmF2EaCvkUN2KWUpydvSBeyrkzyh1mrKPgAA5laLnuyXJHlkuh7sx4zrIkcAAJgWI4XsUsodk/zM4OHeJK8opay5bq31GaO0BQCTtrJiCj+gn1F7sh+eZOfg/x88+O9EhGwAAObCSCG71nppEleBAADAMcZ5x0cAAJhLQjYA9GQKP6AvIRsAABoTsgEAoDEhGwB6MoUf0JeQDQAAjQnZAADQmJANAACNCdkA0JMp/IC+hGwAAGhMyAYAgMaEbADoyRR+QF9CNgAANCZkAwBAY0I2AAA0JmQDQE+m8AP6ErIBAKAxIRsAABoTsgGgJ1P4AX0J2QAA0JiQDQAAjQnZAADQmJANAD2Zwg/oS8gGAIDGhGwAAGhMyAaAnkzhB/QlZAMAQGNCNgAANCZkAwBAY0I2APRkCj+gLyEbAAAaE7IBAKAxIRsAejKFH9CXkA0AAI0J2QAA0JiQDXyNlYMHcvBDb87yxy9zehyOYXYRoK/FSe8AsP0c/OCbcujqdyVJTtn1dVm857dMeI8AmAVfObCcG/cfzL3ucuakd2Xs9GQDX2MYsJNk+e8vneCeADArvnLzcp7/39+bF7/2g7n8qs9PenfGTsgGAGDs3vKez2T58NEkyR+8/RMT3pvxE7IBoCfXKMDJu/nWQ5PehS0lZAMAQGNCNgAANCZkA0BPpvAD+hpLyC6lXFhKWSmlnDeO7QMAwHbWPGSXUh6V5OLW2wUAgGnRNGSXUp6S5B1JdrXcLjB7bj50S97yqbflnZ+9LEdXjk56dwCgqSZ3fCylnJ3kV5M8M8nNSa5PcpcW2wZm0599+tK857r3JUl2n3b7PPyuD57wHsHGTOEH9NWqJ/ul6QL2h5I8MsnszzAOjGQYsJPkr/7x8gnuCQC01ypkfyLJs5I8otb6kUbbBObESvQOAjBbmgwXqbW+rMV2NmtpaTF79py55e1Ook3aUb+N3XTM/y8t7Rj772xxsX8b6je9ZqF2u244/bjHs/Ce+jrRe939xQO91mOytkNdTj/t1OMeb4d9GifzZAMTZ+ZhAGZNk57sSVlePpx9+27ZsvaG37j27r1pgzXZjtTv5CwvHxn77+zw4aMbtqF+02uWanfgwMHjHs/Ce9rIRvVb/Tk8D7+TabKdjr9bDx467vF22KeN7N69M0tLJxeX9WQDAEBjQjYA9GQKP6AvIRsAABoTsoEN6LkDgM0SsgGgp4UFc+EA/QjZwAaECgDYLCEbAAAaG8s82bXWx4xju8Bsclt1poXZRYC+9GQDAEBjQjYwcQvGfQMwY4RsAABoTMgGgJ5M4Qf0JWQDG3ChFwBslpANADRjBhboCNnABpwehyEB8muZghPWJmQDAM2I3NARsgEAoDEhG5g4p5uZFmYXWcPqw9fhDEmEbAAAaE7IBgBOmo5rWJuQDUyc26rD7DD8CzpCNgD0ZAo/oC8hGwA4aau/d/geAh0hGwAAGluc9A4AwLQwhd9aZr/r+siXPpvlK9+Shd13yWkP//4snLJj0rvEFBCyAQDWcfB9f5gj/1STJDv2nJNT7/OICe/RdJq3L6mGiwAbmP1eKuDkzcMY7GHATpIjn//4BPeEaSJkAwDNzEPo5uTM2+w8QjawgfGf3jOvLtNi3kICcPKEbACgIV9EIBGyAQDYAi58BNhibqvOtJi3kNCHfmtYm5ANADQz88PWZ/39jdG8XdMgZAMAQGNCNgBw0lb3Ts58X6URQ/QkZAMbmPmPTOht3k53Q0vzdk2DkA0AAI0J2cAG5qvnARjRrHf2z/r7G6N5OxMkZAMT546PTIt5O90NnDwhGwA4aas7J2f+S7PvWfQkZAMAMHbzdiZIyAYATtrqnuuZH3Y76++PZoRsYOLcVp1pMW8XbkFL83b8CNkAANCYkA1sYL56HoBNmrc/EU680ZOQDQA9zduFW6xh3r5UNDRvx4+QDWxgvv4oApuzOnPO2bBbOCEhGwAAGhOyAYCTp+eanswuAnCc8f9RnPk7xDEz5i0knJwZ/x0ZQUdPQjYAQF8z/h1inObtwsfFVhsqpTwuyc8neWCSpSRXJnlZrfUdrdoAJmG+/igCm/M1d3yc0H7AdtOkJ7uUcmGSdyY5N8kHkrwvybcnubSUclGLNgBg0uatJ461hgj5GkE/I4fsUsrdkvx2kn1JHlprfVKt9QnpQvZXkryylHKPUdsBZpfbqsP0Wp1BZ27YupDdzLxd09CiJ/unk5yW5BW11o8On6y1fjDJryc5PYnebABgCq3+FjGZvWD6tAjZFwyWb1njtTcPlk9s0A4AwNZa1fs6b72xLc3bcKuFUf6xlFIWktyS5NQkO2uty6teX0xyMMmtSXbVWlv9y7wsyfnLy4ezb98tjTa5vkNHDuXNl70m19563eydCpsTw2Nb/TZ2u6M3Hff4wClnNm/j02ccOu7xvW8+dd311W96zVLtblw8ki8vHf3q443+3c6Cjep39OhKDh+97XeytOOU235oBixkJWcc3f/Vx4dzag6ecvoE92hzttPxd+jw0eMulF1a3LGpnz9n5zl5ync+p/VurWv37p1ZWlpMkncnecxmfnbU2UXukG6oyN7VATtJaq2HSylfSnLnJGemG6PdzNLSYvbsaf/hv5Y//cs/zrsXrkl2bklzMGGrP0AOrblWS6tDN0wD/27XcnTjVabO1v9NnA+b+7fy6XwyD7r2qjz8QeeNaX/aGnW4yO0Gy5vXWWfY1bxrxLYm6l53vXdO2Q5fAwEA5tCpR1dy97t+/aR3o7dRe7KHX0HWS58Lq5bNbOVwkbPvcv/88hn/IX/z0SuyvHx4S9qkrcHpHvXr6ZQjt2ZlYTErpzSbTv9r3LpyKDsWTsmp2fiUofpNr1mr3dGs5JaV5dxu4bRJ78qW6FO/lZXk4KHDOX1pfH8vJm1h5UhWMn1DYbbb8Xf06EoOH1nJ0qmb6+ddSPLgf3F+Tls8K3v33rTh+q0cM1xk00Y9GoaDlNYbRDE8x3JgxLYm7r73fkDue+8HbGlxaWc4tEj9ppP6TS+1m27qN93Ub3JGHS7ylXRB+06DixyPM3juTklurbXeOGJbAAAwFUYK2YPZQq5OsiPJ/dZYpQza+Mgo7QAAwDRpMU/2pYPl967x2vC5tzVoBwAApkKLkP3adPNgv6CU8pDhk6WUhyZ5frrZRS5p0A4AAEyFkUN2rfWaJM9Lcvsk7yulvL2UcmmS96abG/uiWusXR20HAACmRYue7NRaL0ny5CTvT/LoJA9LckWSx9daX9+iDQAAmBbNJrSstb41yVtbbQ8AAKZVk55sAADgNkI2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBjCysrK5Peh5NxbZJ7HD26ksOHj2xZo0tL3Q0yl5cPb1mbtKN+0039ppfaTTf1m27qN5rFxR055ZSFJLkuydmb+dlpDdk3Jtk96Z0AAGAu7Ety1mZ+YHE8+zF2n0lyTpL9ST414X0BAGA2fWOSXemy56ZMa082AABsWy58BACAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhMyAYAgMaEbAAAaEzIBgCAxoRsAABoTMgGAIDGhGwAAGhscdI7MC1KKY9L8vNJHphkKcmVSV5Wa33HRHdszpRSnpnkdeus8iu11l84Zv37JfmlJOcl+bokn0ry6iSX1FqPrrH9s5L8XJLvS3LPJNcneVOSX6q1fqXR25g7pZQLk7w2yaNrrVes8fpY61RK2ZHkOUl+Isl9k9yc5F1JfrHW+snR3+HsWq92pZR7JvnHdX78r2ut5636mbOidmMz+H39ZJJnJXlAkh1JPp3kj5K8vNZ666r1HXvbyGbq5/jb/vRk9zD4kHlnknOTfCDJ+5J8e5JLSykXTXDX5tGDBst3JnnDGv99eLhiKeVbk3wwyQ8m+WySS9P9Ubk4awT1Usrtk7w7yfOTHE3y1sHyuUneV0rZPY43NOtKKY9K9zs/0etbUaffTfLbSc5O8o4k1yT5gSRXllIetMb6ZOPa5bbj8e+z9vF4XCeE2o3XIBD9abqa3T/J+5NcluTuSV6S5LJSyhnHrO/Y20Y2W784/rY9PdkbKKXcLd0/sH1Jzqu1fnTw/MOS/GWSV5ZS/rzWet0Ed3OeDA/qZ6/3Oy+lLKT7kLh9kmfWWl8/eH5Purr9cCnlzbXWNx3zYy9Nd6biNUl+otZ6tJSymOT3kzxz8PpPt35Ds6yU8pQkf5Bk1wleH3udBvtwYZK/TfIdtdZ9g+f/bbpj+w9KKd9Wa11p9b5nwUa1Gxgej79ea31Dj82q3Xg9J8l3pwtdTxr+jSyl3CnJnyV5VJIXJfk5x9621Lt+g/Udf9ucnuyN/XSS05K8Yhiwk6TW+sEkv57k9CR6s7fOtyW5vseXmsen+2Ny2fDDI0lqrXuT/NTg4c8Mnx+cQntOkq8ked7wNGmt9fBg/X9O8mOllNu1eRuzrZRydinldelOQ+5Id0pyLVtRp/80WD53+CEx+JnfSRcmHpjkMSf1RmfQJmqX3PYhf2WP7Z4VtRu3CwfLnz32b2St9UvphiAkXa914tjbji4cLPvUL3H8bXtC9sYuGCzfssZrbx4sn7g1uzLfSinnJDkrPf6gZJ261Vr/OskXk5xXSjlz8PS/TLIzybtqrTetWn9/uj8oO5OcfzL7Podemq5n5ENJHpnkEydYb6x1GnywPDLJl5O8Z432HcNfq2/tku5Dfn+SPmMz1W78vpSuXh9Y47Vhje4+WDr2tp/N1C9x/G17housY3A67ZvSjVn6+BqrfHLw2jeXUhacMhm74bf260spF6c7uM9ON5bw9Tn+opBvHiw/mrXVJHdOV9+/6bH+MGh8S5K3ndTez5dPpLtw5/WDU5InWm/cdXpAkoUkV691Edeq9en0ql0p5Y5J7pXuVPJzBxcl3zfJjenGer641vr5Y35E7cas1vrkdV5+2GB57WDp2NtmNlM/x990ELLXd4d0Q0X21lqXV79Yaz1cSvlSuj9EZ6Y7DcP4fHU8drpTW+9Jcl2Sh6a7KOSCUsrjaq23JLnbYN1/OsG2hs/fZbDc7Pqso9b6sp6rjrtO6rpJm6jd8Hh8cLoP2nenCwAPS/LjSZ5cSnlMrbUO1lO7CRl0GL1k8HA4xtqxNyVOUD/H3xQwXGR9w7FJN6+zzi2D5XoXB9HG8I/KG5Pcs9b6PbXW89N9Q78q3ewvLx2ss1HtVtdts+vTxrjrpK7jMzweP5ak1FofX2v97iTnJPlfSe6aboaDIbWbnF9NNwzg+iQvHzzn2Jsea9XP8TcFhOz1DU+RrDcMZGHVkvF5WrpA/cxa64Hhk7XWa9JdMLKS5KJSyqnZuHar67bZ9Wlj3HVS1/F5RZJ7J3lMrfUzwycHx+Zz0p1lekgp5ZGDl9RuAkopL0nywiQHkzx9cGFj4tibCuvUz/E3BQwXWd/+wXLnOuucPlgeWGcdGhiMt776BK99uJRybbo5Xu+XjWu3um6bXZ82xl0ndR2TWuuRJJ85wWs3l1Lele4Cyoekm+9X7bbQYGq2V6Wb/erWJE+ptV5+zCqOvW1so/o5/qaDnuz1fSXdP7Q7Df7BH2fw3J2S3FprvXGL942v9YXB8owkwws+7nqCdVePN9vs+rQx7jqp6+QcezwmardlSim7kvyfdAHtxiRPqLW+fdVqjr1tqmf9NuL42waE7HUMZgu5Ot1csfdbY5WS7nf4ka3cr3lUSjmzlPLqUsr/XusLz8A5g+V1ue0K6m9aY1sL6e6mdSS39YyfcP2BBwyWat3WuOt0dbrTng9Ye3V1PVmllP8yOB5PNLvA8HgczmahdluglHKHdHcJvCDJ55I8elUP9pBjbxvqWz/H33QQsjd26WD5vWu8NnzOlG7jtz/J9yV5ataYq7qUckG6swofGUxbtF7dzk2yJ8kVx8wXenm6Czket/qGM4NehccN9mGt+UI5eWOt02B84hVJ7lxKOXeNNobtOoY374Hpjsenr36hlHLnJN+V5FCS/zt4Wu3GrJSylO738ZB0IencY2+itopjb5vZZP0cf1NAyN7Ya9ONh3pBKeUhwydLKQ9N8vx0/2gvmdC+zY3BWYXXDB5eXEr56oT8pZT75LYaDGcXeXe6q64fX0r58WPW3XPMur9xzPYPJPkf6aZtvGTYW37MuLizkrx69ST+jGwr6jTcziWD2xMP27go3QfL39ZaL2v7tubC7wyWzyulfPvwycEH9u+nu13379Zav5Co3RZ5SbobiHwu3QVx166zrmNv+9lM/Rx/U2BhZcX9UzZSSvmpdP8IDyX5q3RX035HugtHf+TYW9IyPqWUnUn+Isl56b5xXzF46bHp5jP/zVrr845Z/+Hp6rUr3c0UPp/uFrB3SPKaWutFq7Z/xyTvTTcM6NPpJvl/cLoruP8uyb8c3BmLTSqlXJbuDMSja61XrHpt7HUqpfxxuh6ff053KvYeSR6ebrzjebXWj7V6r7Nmg9r9RpLnpjut/Nfp7lj36HRnld6T5IJa683HrK92YzL43V6b7kK1v83aN1BLktRanzH4GcfeNnGS9XP8bXN6snuotV6S5MnprtB9dLrJ3q9I8ngBe+sMbjLznemmM7omXbg+N11dnnpswB6s/4Ekj0g3ef99050++2ySn0jyk2ts/8uD7f1WklPT1fxokl9P8lgBezy2qE4/nO7D6PNJnpTug+KPkjzch8TJGxxzT0/3Af+gdONI/yndWb7vPPYDfrC+2o3Pw3PbTBAPTvd7O9F/SRx728zJ1M/xt83pyQYAgMb0ZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBjQjYAADQmZAMAQGNCNgAANCZkAwBAY0I2AAA0JmQDAEBj/x9QKvm/GgqqQgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 251,
       "width": 364
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot(scores_sample[0])\n",
    "plt.plot(scores_sample[2])\n",
    "plt.plot(scores_sample[74]) # right-loxp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2041, 2662]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i for i,s in enumerate(scores_sample[0]) if s > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'=C,+I=G,=G': 2.317159868479517}\n",
      "{'=T,+I=C,=T': 1.1996800853105838}\n"
     ]
    }
   ],
   "source": [
    "print(mutation_score[2041])\n",
    "print(mutation_score[2662])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2041, 2662の補正はどうなっている？"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=====================\n",
      "control @ 2041\n",
      "defaultdict(<class 'int'>, {'=G': 902, '+I=G': 14, 'N': 51, '+I*GT': 1})\n",
      "=====================\n",
      "sample @ 2041\n",
      "defaultdict(<class 'int'>, {'+I=G': 14, '=G': 356, '+I*GA': 2})\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from collections import Counter\n",
    "\n",
    "count = defaultdict(int)\n",
    "for cs in cssplits_control:\n",
    "    count[cs[2041]] += 1\n",
    "print(\"=====================\")\n",
    "print(f\"control @ {2041}\")\n",
    "print(count)\n",
    "\n",
    "count = defaultdict(int)\n",
    "for cs in cssplits_sample:\n",
    "    count[cs[2041]] += 1\n",
    "print(\"=====================\")\n",
    "print(f\"sample @ {2041}\")\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 1, 1, 1, 1]\n",
      "[1, 1, 1, 1, 1]\n",
      "0.9999999999999998 0.0\n"
     ]
    }
   ],
   "source": [
    "print(count_5mer_sample[2041//5][\"ins\"])\n",
    "print(count_5mer_control[2041//5][\"ins\"])\n",
    "samp = [c / coverage_sample for c in count_5mer_sample[2041//5][\"ins\"]]\n",
    "cont = [c / coverage_control for c in count_5mer_control[2041//5][\"ins\"]]\n",
    "distance = 1 - cosine(samp, cont)\n",
    "_, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "print(distance, pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[12, 11, 10, 6, 10]\n",
      "[5, 20, 14, 19, 9]\n",
      "0.8550649300550162 0.27782700006420435\n"
     ]
    }
   ],
   "source": [
    "print(count_5mer_sample[2662//5][\"ins\"])\n",
    "print(count_5mer_control[2662//5][\"ins\"])\n",
    "samp = [c / coverage_sample for c in count_5mer_sample[2662//5][\"ins\"]]\n",
    "cont = [c / coverage_control for c in count_5mer_control[2662//5][\"ins\"]]\n",
    "distance = 1 - cosine(samp, cont)\n",
    "_, pvalue = stats.ttest_ind(samp, cont, equal_var=False)\n",
    "print(distance, pvalue)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 結果\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### extract_sequence_errors内のdistanceの閾値を0.9とした\n",
    "\n",
    "- 改めてayabe-task1を実行\n",
    "- distanceの閾値を下げたことによってマイナーな変異が補正される恐れがある\n",
    "    - 1%の点変異が正しく抽出できているかを改めて検証する"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 一方で、補正をゆるくするのは根本的な解決にならない\n",
    "\n",
    "+ 1%の点変異が抽出できていれば、0.95->0.9は良いと思う。0.95は厳しすぎる印象。\n",
    "+ しかし、補正をゆるくする→新しい問題が勃発のいたちごっこ\n",
    "    + 2662はdistanceが0.855なので補正されていない\n",
    "    + 2041のinsertionがないのが謎\n",
    "+ 根本の解決は、例えば`insersionのみのアレルを事前に生成する方針`がある"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 次回の課題"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ [x] ~~点変異、欠失、ノックインのデザインで動くかを確認する~~\n",
    "+ [x] ~~Clusteringのときの警告メッセージは無害なので消去する~~\n",
    "+ [ ] right_loxpがいまいちな理由を考察する\n",
    "    +  ~~[x] Ayabe-task1のpreprocessにおいて、329と1280の補正がされていない理由を検討する~~\n",
    "+ [ ] distanceの閾値を下げた状態で、1%点変異が検出できるか\n",
    "+ [ ] `preprocess.correct_sequence_error.replace_atmark`のコードがわかりにくい\n",
    "    + テストを用意してリファクタリングする\n",
    "+ [ ] Insertionのなかにある変異を同定する手法を考案する\n",
    "+ [ ] cis変異の両端が欠失している場合に、Nで置き換えるとtransとなってしまうのをどうするか（`replace_n`）\n",
    "+ [ ] 短いリードの扱いをどうするべきか"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "次回は\n",
    "+ [ ] right_loxpがいまいちな理由を考察する\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a09285d6cbc768c1977f96e8deb5ca1ec0d08675e9573ed6dfd37fd7d91de663"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
